src/server/pkg/hashtree/hashtree.go

```diff
@@ -497,43 +498,110 @@ func (h *hashtree) Finish() (HashTree, error) {
 func (h *hashtree) PutFile(path string, objects []*pfs.Object, size int64) error {
-	return h.putFile(path, objects, nil, size)
+	return h.putFile(path, objects, nil, size, nil, nil, 0)
 }
 
 func (h *hashtree) PutFileOverwrite(path string, objects []*pfs.Object, overwriteIndex *pfs.OverwriteIndex, sizeDelta int64) error {
-	return h.putFile(path, objects, overwriteIndex, sizeDelta)
+	return h.putFile(path, objects, overwriteIndex, sizeDelta, nil, nil, 0)
+}
+
+func (h *hashtree) PutFileSplit(path string, objects []*pfs.Object, size int64, header *pfs.Object, footer *pfs.Object, headerFooterSize int64) error {
+	return h.putFile(path, objects, nil, size, header, footer, headerFooterSize)
 }
 
 // PutFile appends data to a file (and creates the file if it doesn't exist).
-func (h *hashtree) putFile(path string, objects []*pfs.Object, overwriteIndex *pfs.OverwriteIndex, sizeDelta int64) error {
+func (h *hashtree) putFile(path string, objects []*pfs.Object, overwriteIndex *pfs.OverwriteIndex, sizeDelta int64, header *pfs.Object, footer *pfs.Object, headerFooterSize int64) error {
 	path = clean(path)
 
 	// Detect any path conflicts before modifying 'h'
 	if err := h.visit(path, nop); err != nil {
 		return err
 	}
 
-	// Get/Create file node to which we'll append 'objects'
+	if len(objects) > 0 {
+		// Get/Create file node to which we'll append 'objects'
+		node, ok := h.fs[path]
+		if !ok {
+			node = &NodeProto{
+				Name:     base(path),
+				FileNode: &FileNodeProto{},
+			}
+			h.fs[path] = node
+		} else if node.nodetype() != file {
+			return errorf(PathConflict, "could not put file at \"%s\"; a file of "+
+				"type %s is already there", path, node.nodetype().tostring())
+		}
+
+		// Append new objects.  Remove existing objects if overwriting.
+		if overwriteIndex != nil && overwriteIndex.Index <= int64(len(node.FileNode.Objects)) {
+			node.FileNode.Objects = node.FileNode.Objects[:overwriteIndex.Index]
+		}
+		node.SubtreeSize += sizeDelta + headerFooterSize
+		node.FileNode.Objects = append(node.FileNode.Objects, objects...)
+		h.changed[path] = true
+		// Add 'path' to parent (if it's new) & mark nodes as 'changed' back to root
+		err := h.visit(path, func(node *NodeProto, parent, child string) error {
+			if node == nil {
+				node = &NodeProto{
+					Name:    base(parent),
+					DirNode: &DirectoryNodeProto{},
+				}
+				if parent == filepath.Dir(path) {
+					node.SubtreeSize = headerFooterSize
+					emptyObject := pfs.Object{}
+					if header != nil {
+						if *header == emptyObject {
+							node.DirNode.Header = nil
+						} else {
+							node.DirNode.Header = header
+						}
+					}
+					if footer != nil {
+						if *footer == emptyObject {
+							node.DirNode.Footer = nil
+						} else {
+							node.DirNode.Footer = footer
+						}
+					}
+				}
+				h.fs[parent] = node
+			}
+			insertStr(&node.DirNode.Children, child)
+			node.SubtreeSize += sizeDelta
+			h.changed[parent] = true
+			return nil
+		})
+		return err
```

Comment:
Put the return on the line above, there's no reason to store an error in a variable if the next thing you're doing is returning it.